thrifty-runtime/src/main/java/com/microsoft/thrifty/service/ClientBase.java

```diff
@@ -67,7 +67,11 @@ protected final Object execute(MethodCall<?> methodCall) throws Exception {
             throw new IllegalStateException("Cannot write to a closed service client");
         }
 
-        return invokeRequest(methodCall);
+        try {
+            return invokeRequest(methodCall);
+        } catch (UndeclaredServerException e) {
+            throw e.thriftException;
```

Comment:
This maintains the old behavior of `handleExceptionMessage` which just threw the exception. I'm not sure if that is really what we want for synchronous clients. We have the exception wrapping explicitly so that we can pass it on to `fail()` for asynchronous call. That way the client doesn't get closed and we can continue to execute different calls. 
Should be expose `UndeclaredServerException` as public API so that a synchronous call has to catch it? Throwing a runtime exception is pretty similar to closing the async client and catching `ThriftException` is probaby not a good idea because in other cases you want it to go through.